R/tmb-methods.R

```diff
@@ -429,3 +429,102 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+#' @describeIn mmrm_tmb_methods Simulate from an mmrm model object
+#' @importFrom stats simulate
+#'
+#' @param object (`mmrm`)\cr a fitted 'mmrm' object
+#' @param nsim (`integer`)\cr number of replications to simulate.
+#' @param seed unused argument from simulate generic.
+#' @param newdata (`data.frame`)\cr new data.frame for which to simulate values.
+#' @param method (`string`)\cr one of "conditional" or "marginal". If "conditional",
+#' simulated values are sampled given the estimated covariance matrix of `object`. If "marginal",
+#' variance of the estimated covariance matrix is taken into account.
+#'
+#' @return A `data.frame` of dimension `[n, m]` where n is the number of rows in `newdata`,
+#' and m is the number `nsim` of simulated responses.
+#' @exportS3Method
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                              seed = NULL, ...,
+                              newdata = NULL,
+                              method = c("conditional", "marginal")) {
+  method <- match.arg(method)
+
+  if (is.null(newdata)) {
+    newdata <- object$data
+  }
+
+  # Ensure new data has the same levels as original data.
+  full_frame <- model.frame(
+    object,
+    data = newdata,
+    include = c("subject_var", "visit_var", "group_var", "response_var"),
+    na.action = "na.pass"
+  )
+  tmb_data <- h_mmrm_tmb_data(
+    object$formula_parts, full_frame,
+    weights = rep(1, nrow(full_frame)),
+    reml = TRUE,
+    singular = "keep",
+    drop_visit_levels = FALSE,
+    allow_na_response = TRUE,
+    drop_levels = FALSE
+  )
+
+  if (method == "conditional") {
+    mu <- h_get_prediction(tmb_data, object$theta_est, object$beta_est, object$beta_vcov)
+    ret <- as.data.frame(h_get_sim_per_subj(mu, tmb_data$n_subjects, nsim))
+  } else if (method == "marginal") {
+    theta_chol <- t(chol(object$theta_vcov))
+    ntheta <- length(object$theta_est)
+    ret <- as.data.frame(
+      sapply(seq_len(nsim), function(x) {
+        newtheta <- theta_chol %*% matrix(rnorm(ntheta), nrow = 1) + object$theta_est
```

Comment:
hm this cannot work with just one row but needs one column because the cholesky matrix will have more than one column?
```suggestion
        newtheta <- theta_chol %*% matrix(rnorm(ntheta), ncol = 1) + object$theta_est
```